lib/Extension/LanguageServerCodeTransform/LspCommand/GenerateDecoratorCommand.php

```diff
@@ -0,0 +1,126 @@
+use Phpactor\Extension\LanguageServerBridge\Converter\TextEditConverter;
+use Phpactor\LanguageServerProtocol\Command;
+use Phpactor\LanguageServerProtocol\WorkspaceEdit;
+use Phpactor\LanguageServer\Core\Server\ClientApi;
+use Phpactor\CodeTransform\Domain\SourceCode;
+use Phpactor\LanguageServer\Core\Workspace\Workspace;
+use Phpactor\TextDocument\TextEdits;
+use Phpactor\WorseReflection\Reflector;
+use Phpactor\CodeBuilder\Domain\Builder\MethodBuilder;
+use Phpactor\WorseReflection\Core\Reflection\ReflectionMethod;
+use Phpactor\CodeBuilder\Domain\Prototype\Visibility;
+
+class GenerateDecoratorCommand extends Command
+{
+    public const NAME = 'generate_decorator';
+
+    private ClientApi $clientApi;
+
+    private Workspace $workspace;
+
+    private Reflector $reflector;
+
+    private Updater $updater;
+
+    public function __construct(
+        ClientApi $clientApi,
+        Workspace $workspace,
+        Reflector $reflector,
+        Updater $updater
+    ) {
+        $this->clientApi = $clientApi;
+        $this->workspace = $workspace;
+        $this->reflector = $reflector;
+        $this->updater = $updater;
+    }
+
+    /**
+     * @return Promise<ApplyWorkspaceEditResponse>
+     */
+    public function __invoke(string $uri, string $interfaceFQN): Promise
+    {
+        $textDocument = $this->workspace->get($uri);
+        $source = SourceCode::fromStringAndPath($textDocument->text, $textDocument->uri);
+
+        $textEdits = $this->getTextEdits($source, $interfaceFQN);
+
+        return $this->clientApi->workspace()->applyEdit(new WorkspaceEdit([
+            $uri => TextEditConverter::toLspTextEdits($textEdits, $textDocument->text)
+        ]), 'Generate decoration');
+    }
+
+    private function getTextEdits(SourceCode $source, string $interfaceFQN): TextEdits
+    {
+        $class = $this->reflector->reflectClassesIn($source)->classes()->first();
+
+        $builder = SourceCodeBuilder::create();
+        $builder->namespace($class->name()->namespace());
+        $classBuilder = $builder->class($class->name()->short());
+
+        $classBuilder->property('inner')->visibility(Visibility::PRIVATE)->type($interfaceFQN);
+
+        $constructor = $classBuilder->method('__construct');
+        $constructor->parameter('inner')->type($interfaceFQN);
+        $constructor->body()->line('$this->inner = $inner');
+
+        $interface = $this->reflector->reflectInterface($interfaceFQN);
+        foreach ($interface->methods() as $interfaceMethod) {
+            $method = $classBuilder->method($interfaceMethod->name());
+
+            $method->returnType($interfaceMethod->returnType());
+            $method->visibility($interfaceMethod->visibility());
+
+            $this->attachParameters($method, $interfaceMethod);
+
+            $method->body()->line($this->generateMethodBody($interfaceMethod));
+        }
+
+        return $this->updater->textEditsFor($builder->build(), Code::fromString((string) $source));
```

Comment:
the "logic" shouldn't be in this class (think of it like an MVC controller), it should be in a plain PHP class in CodeTransform such as:

https://github.com/phpactor/phpactor/blob/9f20e64fcc828851157fc4e1896a1278b73f3f97/lib/CodeTransform/Adapter/WorseReflection/Refactor/WorseExtractConstant.php#L24

(you could add an interface as with that example too)

the class doesn't have to conform to any specific specification, just has to do this job (i.e. it returns the `TestEdits` collection)

you can delegate to that class from here,

Reply:
for testing _the logic_ you can look at any of the examples in CodeTransform, e.g. for the above:

https://github.com/phpactor/phpactor/blob/c102bef07614c48a53fb552a1fd56663b5a1949d/lib/CodeTransform/Tests/Adapter/WorseReflection/Refactor/WorseExtractConstantTest.php